Framework/PythonInterface/plugins/algorithms/DeltaPDF3D.py

```diff
@@ -0,0 +1,251 @@
+
+        if self.getProperty("SpaceGroup").value:
+            space_group=self.getProperty("SpaceGroup").value
+            try:
+                if not SpaceGroupFactory.isSubscribedNumber(int(space_group)):
+                    issues["SpaceGroup"] = 'Space group number is not valid'
+            except ValueError:
+                if not SpaceGroupFactory.isSubscribedSymbol(space_group):
+                    issues["SpaceGroup"] = 'Space group name is not valid'
+
+        sphereMin = self.getProperty("SphereMin").value
+        if len(sphereMin) != 1 and len(sphereMin) != 3:
+            issues["SphereMin"] = 'Must provide 1 or 3 widths'
+
+        sphereMax = self.getProperty("SphereMax").value
+        if len(sphereMax) != 1 and len(sphereMax) != 3:
+            issues["SphereMax"] = 'Must provide 1 or 3 widths'
+
+        return issues
+
+    def PyExec(self):
+        progress = Progress(self, 0.0, 1.0, 5)
+        inWS = self.getProperty("InputWorkspace").value
+        signal = inWS.getSignalArray().copy()
+
+        dimX=inWS.getXDimension()
+        dimY=inWS.getYDimension()
+        dimZ=inWS.getZDimension()
+        X=np.linspace(dimX.getMinimum(),dimX.getMaximum(),dimX.getNBins()+1)
+        Y=np.linspace(dimY.getMinimum(),dimY.getMaximum(),dimY.getNBins()+1)
+        Z=np.linspace(dimZ.getMinimum(),dimZ.getMaximum(),dimZ.getNBins()+1)
+        Xs, Ys, Zs = np.mgrid[(X[0]+X[1])/2:(X[-1]+X[-2])/2:dimX.getNBins()*1j,
+                              (Y[0]+Y[1])/2:(Y[-1]+Y[-2])/2:dimY.getNBins()*1j,
+                              (Z[0]+Z[1])/2:(Z[-1]+Z[-2])/2:dimZ.getNBins()*1j]
+
+        if self.getProperty("RemoveReflections").value:
+            progress.report("Removing Reflections")
+            width = self.getProperty("Width").value
+            if len(width)==1:
+                width = np.repeat(width, 3)
+            cut_shape = self.getProperty("Shape").value
+            space_group = self.getProperty("SpaceGroup").value
+            if space_group:
+                check_space_group = True
+                try:
+                    space_group=SpaceGroupFactory.subscribedSpaceGroupSymbols(int(space_group))[0]
+                except ValueError:
+                    pass
+                logger.information('Using space group: '+space_group)
+                sg=SpaceGroupFactory.createSpaceGroup(space_group)
+            else:
+                check_space_group = False
+            for h in range(int(np.ceil(dimX.getMinimum())), int(np.floor(dimX.getMaximum()))+1):
+                for k in range(int(np.ceil(dimY.getMinimum())), int(np.floor(dimY.getMaximum()))+1):
+                    for l in range(int(np.ceil(dimZ.getMinimum())), int(np.floor(dimZ.getMaximum()))+1):
+                        if not check_space_group or sg.isAllowedReflection([h,k,l]):
+                            if cut_shape == 'cube':
+                                x_min=np.searchsorted(X,h-width[0])
```

Comment:
Is `searchsorted` faster than calculating the index by hand? You know that MDHistoWorkspaces have a regular grid?